### DISCUSSION TITLE
Why are rocks are configured before being loaded?

### DISCUSSION THREAD

**Discussion by austinliuigi** [nvim-neorocks/rocks-lazy.nvim#7] [Q&A] [answered]
I was reading over the source code and was just curious why `config_hook` is called *before* loading the plugin rather than after. Since it's called before, it needs to `:packadd!` before calling `rocks_config.configure()` which seems redundant, as `lz.n` calls `:packadd` during the load step regardless.

**Comment by mrcjkb** [comment, marked as answer]
Hey :wave: 

This is because it's more in line with Neovim's built-in [`:h initialization-startup`](https://neovim.io/doc/user/starting.html#initialization).
Neovim sources your `init.lua` first, and then loads plugins later.

Typically, there are two types of plugins:

- Those that allow optional configuration before **automatic** initialization
- Those that force you to configure and initialize at the same time with a `setup` function

Requiring configuration after initialization would be a really bad idea (and I haven't seen anyone do it), because it would not work with Neovim's built-in loading mechanism (and there probably isn't a use case for it).

rocks-lazy.nvim's  loading mechanisms occur after Neovim's initialization startup.
Calling `packadd!` with a bang (`!`) makes the Lua API[^1] of non-luarocks plugins that are managed by rocks-git.nvim available on the runtimepath, without sourcing any plugin scripts.

[^1]: which may be needed for configuration

Then, after configuration, `packadd` (without a bang) sources the plugin scripts.

You're right, the `packadd!` call is unnecessary for luarocks packages, because their Lua API is on the `package.path`/`package.cpath` already. But it's needed for non-luarocks packages managed by rocks-git.nvim. 

See also: https://github.com/nvim-neorocks/rocks-lazy.nvim/issues/4

**Reply by austinliuigi** [reply]
Got it, thanks for the thorough response!
